src/topo.rs

```diff
@@ -0,0 +1,565 @@
+                    let tmp = source_que.iter().last().copied().unwrap();
+                    source_que.remove(&tmp);
+
+                    Some(tmp.1)
+                } else {
+                    // paper says "Then we are done"
+                    None
+                };
+
+                // We don't ever use this number again if `s` is None but
+                // we can't return early
+                s_indegree = s.map_or(0, |id| {
+                    self.nodes.get(&id).map(|n| n.in_degree()).unwrap_or(0)
+                });
+            }
+
+            if t_outdegree == 0 {
+                for idx_node_pair in self.nodes_with_edges_outgoing(some_t) {
+                    target_que.insert(idx_node_pair);
+                }
+                target_stack.push(some_t);
+
+                t = if !target_que.is_empty() {
+                    // TODO:
+                    // use `target_que.pop_first()`
+                    let tmp = target_que.iter().next().copied().unwrap();
+                    target_que.remove(&tmp);
+
+                    Some(tmp.1)
+                } else {
+                    // paper says "Then we are done"
+                    None
+                };
+
+                // We don't want to return early and since we don't care about
+                // this number if `t` is None
+                t_outdegree = t.map_or(0, |id| {
+                    self.nodes.get(&id).map(|n| n.out_degree()).unwrap_or(0)
+                });
+            }
+        }
+
+        // Relabel the nodes...
+
+        if s.map_or(true, |s| s == to) {
+            s = self.previous(to);
+        }
+        if t.map_or(true, |t| t == from) {
+            t = self.next(from);
+        }
+
+        for s_first in source_stack.drain(..).rev() {
+            self.insert_after(s_first, s)?;
+            s = Some(s_first);
+        }
+
+        for t_last in target_stack.drain(..).rev() {
+            self.insert_before(t_last, t)?;
+            t = Some(t_last);
+        }
+
+        let s = s.ok_or(TopoError::NotFound)?;
+        let t = t.ok_or(TopoError::NotFound)?;
+        self.add_edge_to_graph(s, t)
+    }
+
+    fn add_edge_to_graph(&mut self, from: Id, to: Id) -> TopoResult<()> {
+        self.nodes
+            .get_mut(&from)
+            .ok_or(TopoError::NotFound)?
+            .add_outgoing_edge(to);
+        self.nodes
+            .get_mut(&to)
+            .ok_or(TopoError::NotFound)?
+            .add_incoming_edge(from);
+        Ok(())
+    }
+
+    fn is_before(&self, from: Id, to: Id) -> TopoResult<bool> {
+        let s = self.nodes.get_index_of(&from).ok_or(TopoError::NotFound)?;
+        let t = self.nodes.get_index_of(&to).ok_or(TopoError::NotFound)?;
+        Ok(s < t)
+    }
+
+    fn previous(&self, node: Id) -> Option<Id> {
+        self.nodes
+            .get_index_of(&node)
+            .and_then(|idx| idx.checked_sub(1))
+            .and_then(|idx| self.nodes.get_index(idx).map(|(k, _)| *k))
```

Comment:
Probably more readable when written with the try operator, e.g.:

```suggestion
        let idx = self.nodes.get_index_of(&node)?.checked_sub(1)?;
        Some(self.nodes.get_index(idx)?.0)
```